### PR TITLE
Update gradient logic for Qiskit Rust circuit data implementation

### LINKED ISSUE #181: Nightly test against Qiskit main failing
Some gradient tests are failing due to changes in QuantumCircuit

```
      File "/home/runner/work/qiskit-algorithms/qiskit-algorithms/qiskit_algorithms/gradients/reverse/derive_circuit.py", line 135, in derive_circuit
    if len(circuit._parameter_table[parameter]) > 1:

    AttributeError: 'QuantumCircuit' object has no attribute '_parameter_table'
```

The change came from Qiskit/qiskit#12459 which is moving circuit logic into Rust.

### PR DESCRIPTION
**Pull request by woodsp-ibm**
### Summary

Fixes #181

This updates internals of the gradients which more directly manipulated circuit data the way it was when it was managed in Python. The Rust implementation is changing things to be more performant but things are also different so this changes the logic to try accommodate as things are currently done, which is what is in Qiskit stable releases at present, and the new Rust implementation which is presently in main branch of Qsikit repo. #181 notes the aspect of CI which is failing is when we test against main which was done to monitor for any problems here due to changes.

This has partially fixed things but there is another aspect, that I failed to notice/note in #181 that also occurs
```
      File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/qiskit/circuit/parameterexpression.py", line 418, in __float__
    raise TypeError(

    TypeError: ParameterExpression with unbound parameters (dict_keys([Parameter(__gθ0)])) cannot be cast to a float.
```
which has not been addressed at present.

### CODE REVIEW

**Comment by woodsp-ibm** [comment]
While this would be mergeable, as the tests against main, though they fail do not prevent things from being merged. I labelled it onHold as it would be better to have the remaining problem, as noted above, sorted so things pass completely against main.

File: qiskit_algorithms/gradients/utils.py
Comment (ElePT) [inline qiskit_algorithms/gradients/utils.py:146]: Not ideal but I agree that it's the most straightforward workaround so far.

**Review by ElePT — APPROVED** [review]
Thanks a lot @woodsp-ibm for keeping the code updated. I was planning to eventually look into it but you got there faster :)

File: qiskit_algorithms/gradients/utils.py
Comment (ElePT) [reply qiskit_algorithms/gradients/utils.py:146]: Ok I just talked to @Cryoris and I agree that the dagcircuit approach is nicer. Let's wait to see if his implementation makes it :)

Comment (woodsp-ibm) [reply qiskit_algorithms/gradients/utils.py:146]: There is a PR against the branch in my fork to alter this for a DAG if you want to take a look https://github.com/woodsp-ibm/qiskit-algorithms/pull/1 If you and he feel its better with that as a replacement I am happy to have that instead. In the case of this PR I was thinking to leave it open until Qiskit 1.2 as I believed there was a possibility of that the problem may be alleviated. But either way perhaps the DAG approach is just better and we should go for that anyway.

**Comment by woodsp-ibm** [comment]
The last change, to file gradients/utils.py, is not needed with this PR Qiskit/qiskit/pull/12794, so once this merges to Qiskst main I will remove the workaround from that (i,e the data copy)

**Comment by Cryoris** [comment]
https://github.com/Qiskit/qiskit/pull/12794 (Jake's PR) fixes parts of this PR, as it removes the need for copying `circuit.data`. However, it creates new issues due to the fact that `derive_circuit` uses the private `_parameter_table` attribute. To avoid such breakage in future, https://github.com/woodsp-ibm/qiskit-algorithms/pull/2 rewrites `derive_circuit` to only use public API. With that PR and https://github.com/Qiskit/qiskit/pull/12794 all tests pass as expected.

So we can either
1. Merge this as is and merge the public API PR as follow-up fix once Jake's PR is released
2. Wait until Jake's PR is released in 1.2 and merge this PR with the PR that changes to using only public API

With Qiskit 1.2 imminent maybe option 2 is better?

File: qiskit_algorithms/gradients/utils.py
Comment (Cryoris) [reply qiskit_algorithms/gradients/utils.py:146]: Using DAGs is slower unfortunately, so that's off the table for now 🙂

**Comment by woodsp-ibm** [comment]
@Cryoris Well it seems that things need a little fixup having merged your PR. A couple of copyright dates and mypy typing by the looks.

@ElePT It seems with main that more stuff is broke - maybe with 1.2 as well. This PR Qiskit/qiskit#12824 changed base types so  computeuncompute now fails e.g.

```
test.time_evolvers.test_pvqd.TestPVQD.test_zero_parameters
----------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/runner/work/qiskit-algorithms/qiskit-algorithms/test/time_evolvers/test_pvqd.py", line 216, in test_zero_parameters
    fidelity_primitive = ComputeUncompute(sampler)

      File "/home/runner/work/qiskit-algorithms/qiskit-algorithms/qiskit_algorithms/state_fidelities/compute_uncompute.py", line 83, in __init__
    raise ValueError(

    ValueError: The sampler should be an instance of BaseSampler, but got <class 'qiskit.primitives.sampler.Sampler'>
```

**Comment by Cryoris** [comment]
Copyright & mypy done in 0c99ccb

**Comment by woodsp-ibm** [comment]
Sampler base class change reverted in Qiskit/qiskit#12871

**Comment by woodsp-ibm** [comment]
With Qiskit/qiskit#12794 having been merged earlier today I re-ran the CI and it all passes now. So it looks like this is good to go.

**Review by Cryoris — APPROVED** [review]
LGTM but having touched the code maybe @ElePT should approve too 😄

**Pull request merged**
